design/app-to-app.md

```diff
@@ -0,0 +1,87 @@
+
+`SharingMessage` is itself relatively straightforward, but contains enough information to allow the recipient (callback listener) to cast it to the appropriate type (e.g., `PhotoSharingMessage`). We envision multiple message types and don't want to add a registration api for each.
+
+```ts
+class SharingMessage {
+    messageType: Readonly<string>;
+    sessionId: Readonly<string>;
+    userId: Readonly<Guid>;
+}
+
+class PhotoSharingMessage extends SharingMessage {
+    photoType: Readonly<string>; // Supported values: "jpg" and "png"
+    data: Readonly<Buffer>; // Raw bytes of an image file in the photoType format above
+}
+```
+
+`SharingCallbacks` are a relatively straightforward callback registration mechanism, with an optional mechanism to allow the callback registrant to stop processing of messages and construction of payloads in cases where a particular message would not be acted on.
+
+```ts
+
+// Fires before the MessageReceivedCallback. When the implementation
+// returns false, may short-circuit additional processing on the message.
+// Note that this callback is optional but may be useful if the
+// application is unable or uninterested in acting on all messages.
+interface IsMessageDesiredCallback { 
+    (sessionId: string, userId: Guid) : bool
+}
+
+interface MessageReceivedCallback {
+    (sharingMessage: SharingMessage) : void
+}
+
+class SharingCallbacks {
+    // Note that this callback is optional (See IsMessageDesiredCallback).
+    // Leave null to process all messages 
+    public isMessageDesired : IsMessageDesiredCallback;
+    public messageReceived : MessageReceivedCallback;
+}
+```
+
+Callbacks are registered (and deregistered) via new functions on the context described below. Optionally, callbacks may be scoped to a particular session id or user id. When no session or user id is provided, sharing data will be accepted from any session and user.
+
+For the sake of simplicity, only one callback may be registered for a given ```<MessageType,UserId>``` on a particular session context at a time. This restriction may allow for a more simple bookeeping implementation for registered callbacks within the the sharing implementation on both the app side and the client runtime See [Implementation Suggestions](#Implementation-Suggestions)
+
+```ts
+// Modifications to
+// https://github.com/microsoft/mixed-reality-extension-sdk/blob/master/packages/sdk/src/core/context.ts
+
+/**
+ * Registers a set of callbacks invoked at various points in the processing
+ * of a sharing message. Only one callback may be registered at a time for a 
+ * given context and user.
+ * 
+ * Use ZeroGuid to register for all users.
+ * 
+ * Returns true if the callback was successfully registered.
+ * @event
+ */
+public onSharingMessage(userId: Guid, callbacks: SharingCallbacks): bool {
```

Comment:
onSharingMessage does not take a type, it looks like it registers for all SharingMessages.  In a future where there are multiple types of sharing (photo, video, GLTF, text, etc), how will an MRE express the types that it supports?

I see 3 obvious techniques:

  1.  Realtime filtering:  the MRE could receive the message type in IsMessageDesired and reject unsupported sharing types. 

   2.  Runtime type registration:  onSharingMessage takes a specific type or maybe a list of supported sharing message types.

   3.  Manifested type registration:  the manifest.json lists the supported types.

I am just not sure that #1 is ideal.  If the host wants to generate a list of viable targets for a given sharing action, it would need to broadcast this message to every running MRE.  But sending a sharing message to an MRE the user did not explicitly select feels like a privacy risk, we'd have to be very careful with the metadata.  This is also unlikely to be compatible with OS-provided sharing constructs, which expect registration ahead of time.

#2 invites seems well aligned with supporting future scenarios.  For instance, messaging apps commonly register specific sharing targets/shortcuts for recent conversations in addition to their global sharing target.  It seems like a runtime registration design would be needed to support dynamic/multiple registrations of that sort in the future.

Reply:
Very good catch. Let me add messageType to the registration. This was an oversight and it showed up as i was writing the proposed sample data flow. Let's go with 2. I don't think it precludes us from adding 3 down the road, e.g., a manifest might help the host implement disambiguation of the share target if needed (the question of whether each session gets its own registration on the host or not could be aided by manifest).